02-sql-aggregation.md

```diff
@@ -114,32 +114,121 @@ think of views is as a table, that can read, aggregate, and filter information
 from several places before showing it to you.
 
 Creating a view from a query requires to add `CREATE VIEW viewname AS`
-before the query itself. For example, if we want to save the query giving
-the number of individuals in a view, we can write
+before the query itself. For example, imagine that my project only covers
+the data gathered during the summer (May - September) of 2000.  That 
+query would look like: 
 
-    CREATE VIEW species_count AS
-    SELECT species_id, COUNT(*)
+    SELECT *
     FROM surveys
-    GROUP BY species_id;
+    WHERE year == 2000 AND (month > 4 AND month < 10)
 
-Now, we will be able to access these results with a much shorter notation:
+But we don't want to have to type that every time we want to ask a 
+question about that particular subset of data!  Let's create a view: 
 
+    CREATE VIEW summer_2000 AS
     SELECT *
-    FROM species_count;
-
-Assuming we do not need this view anymore, we can remove it from the database
-almost as we would a table:
-
-    DROP VIEW species_count;
+    FROM surveys
+    WHERE year == 2000 AND (month > 4 AND month < 10)
 
 You can also add a view using *Create View* in the *View* menu and see the
 results in the *Views* tab just like a table
 
+Now, we will be able to access these results with a much shorter notation:
+
+    SELECT *
+    FROM summer_2000;
+
 > ### Challenge
 >
 > Write a query that returns the number of each species
 caught in each year sorted from most often caught species to the least
 occurring ones within each year starting from the most recent records. Save
 this query as a `VIEW`.
 
+## Null values
+
+Using the view we created in the previous section (`summer_2000`), let's 
+talk about null values.  Missing values in SQL are identified with the 
+special NULL value.  Scroll through our `summer_2000` view.  It should be 
+easy to find several records with missing values.  How do you think we 
+could filter to find these rows?  
+
+You might want to try a query like: 
+
+    SELECT *
+    FROM summer_2000
+    WHERE species_id == NULL
+
+But this won't work, because the NULL value isn't a text string, it's 
+a special value.  Instead we need the following syntax: 
+
+    SELECT *
+    FROM summer_2000
+    WHERE species_id IS NULL
+
+If we wanted to use all the records where `species_id` is NOT null, we 
+would add the `NOT` keyword to our query.  
+
+    SELECT *
+    FROM summer_2000
+    WHERE species_id IS NOT NULL
+
+There are many hidden "gotchas" with NULL values.  If we restrict our 
+query to the "PE" species, this will be easier to see: 
```

Comment:
Sure - wikipedia's [Three-value logic](https://en.wikipedia.org/wiki/Three-valued_logic) article is fairly thorough (including a portion on SQL's interpretation at the end). A quick summary assuming people are already familiar with the Boolean logic operators (AND, OR, NOT) might go something like:

TRUE and FALSE are "definite" values. Think of NULL not so much as missing data, but as "unknown" data (it has a value, we just don't know which value it is). Now consider the logic operators: AND is true when both its operands are TRUE. Consider the case of TRUE AND NULL: one operand is TRUE, the other might be TRUE or FALSE. The result depends on the unknown NULL value, ergo the result is also NULL. But what about FALSE AND NULL? Since one operand is FALSE; it doesn't matter what the other is: the result must be FALSE.

Likewise for OR, FALSE OR NULL is NULL, TRUE OR NULL is TRUE. What about NOT? NOT NULL is simply NULL for the same reason: the result would depend on what the NULL is.

This also explains what happens when testing for equality with NULLs. Consider an INTEGER field like "year" and the search condition "WHERE year = 2000". When "year" is 2000, the result of the test is TRUE so WHERE includes the row in the result. If "year" is 2001, the result of the test is FALSE, so WHERE excludes the row. If "year" is NULL it represents any possible year, so the result of the test is NULL. WHERE excludes the row in the last case because WHERE's condition must evaluate strictly to TRUE (not FALSE or NULL) in order for the row to be included.